cpp/ql/test/library-tests/members/this/test.cpp

```diff
@@ -0,0 +1,46 @@
+
+int global;
+
+class C {
+  int x;
+
+public:
+
+  void f1() {
+    // Implicit dereference of `this.`
+    x++;
+  }
+
+  void f2() {
+    // Explicit dereference of `this.`
+    this->x++;
+  }
+
+  int f3() const {
+    // We expect the type of `this` to be const-qualified.
+    return x;
+  }
+
+  int f4() volatile {
+    // We expect the type of `this` to be volatile-qualified.
+    return x;
+  }
+
+  int f5() const volatile {
+    // We expect the type of `this` to be qualified as both const and volatile.
+    return x;
+  }
+
+  void f6() {
+    // No use of `this`, but we still expect to be able to get its type.
+    global++;
+  }
+};
```

Comment:
At least two more test cases:
```
float f7() const & {
  // We expect the type of `this` to be const-qualified
  return x;
}

float f8() && {
  // We expect the type of `this` to be unqualified
  return x;
}
```
Neither ref qualifier should affect the type of `this`, but I think it's worth checking.

Reply:
Thanks, I'll add those.

Reply:
Cool.